src/map.c

```diff
@@ -1,754 +1,612 @@
-        return;
-
-    /* Find out the identity */
-    map_node_t *parent = rb_parent(node), *grandparent = rb_parent(parent),
-               *uncle;
-
-    if (!rb_parent(parent))
-        return;
-
-    /* Find out the uncle */
-    if (grandparent->left == parent)
-        uncle = grandparent->right;
-    else
-        uncle = grandparent->left;
-
-    if (rb_is_red(uncle)) {
-        /* If the uncle is red, change color of parent and uncle to black */
-        rb_set_color(uncle, RB_BLACK);
-        rb_set_color(parent, RB_BLACK);
-
-        /* Change color of grandparent to red. */
-        rb_set_color(grandparent, RB_RED);
-
-        /* Call this on the grandparent */
-        map_fix_colors(obj, grandparent);
+    assert(nodep->node == node);
+    pathp--;
+    if (pathp->node != node) {
+        /* Swap node with its successor. */
+        bool tred = rbtn_red_get(pathp->node);
+        rbtn_color_set(pathp->node, rbtn_red_get(node));
+        rbtn_left_set(pathp->node, rbtn_left_get(node));
+        /* If node's successor is its right child, the following code
+         * will do the wrong thing for the right child pointer.
+         * However, it doesn't matter, because the pointer will be
+         * properly set when the successor is pruned.
+         */
+        rbtn_right_set(pathp->node, rbtn_right_get(node));
+        rbtn_color_set(node, tred);
+        /* The pruned leaf node's child pointers are never accessed
+         * again, so don't bother setting them to nil.
+         */
+        nodep->node = pathp->node;
+        pathp->node = node;
+        if (nodep == path) {
+            rbtree->root = nodep->node;
+        } else {
+            if (nodep[-1].cmp < 0) {
+                rbtn_left_set(nodep[-1].node, nodep->node);
+            } else {
+                rbtn_right_set(nodep[-1].node, nodep->node);
+            }
+        }
     } else {
-        /* If the uncle is black. */
-        if (parent == grandparent->left && node == parent->left)
-            map_l_l(obj, node, parent, grandparent, uncle);
-        else if (parent == grandparent->left && node == parent->right)
-            map_l_r(obj, node, parent, grandparent, uncle);
-        else if (parent == grandparent->right && node == parent->left)
-            map_r_l(obj, node, parent, grandparent, uncle);
-        else if (parent == grandparent->right && node == parent->right)
-            map_r_r(obj, node, parent, grandparent, uncle);
+        map_node_t *left = rbtn_left_get(node);
+        if (left) {
+            /* node has no successor, but it has a left child.
+             * Splice node out, without losing the left child.
+             */
+            assert(!rbtn_red_get(node));
+            assert(rbtn_red_get(left));
+            rbtn_black_set(left);
+            if (pathp == path) {
+                rbtree->root = left;
+                /* Nothing to summarize -- the subtree rooted at the
```

Comment:
See if "Nothing to summarize" makes sense or not. We shall ignore the handling of summary.